File: node/drizzle/01_connect.js
// 01_connect.js — Connecting to CUBRID with Drizzle ORM.
//
// Demonstrates:
// - Creating a drizzle instance with cubrid-client
// - Running raw SQL with sql template tag
// - Connection metadata (version, database, user)

import { createClient } from "cubrid-client";
import { drizzle } from "drizzle-cubrid";
import { sql } from "drizzle-orm";

const DB_CONFIG = {
  host: "localhost",
  port: 33000,
  database: "testdb",
  user: "dba",
  password: "",
};

async function basicConnection() {
  console.log("=== Basic Connection ===");
  const client = createClient(DB_CONFIG);
  const db = drizzle(client);

  const rows = await db.execute(sql`SELECT 1 + 1 AS result`);
  console.log(`1 + 1 = ${rows[0].result}`);

  await client.close();
}

async function connectionInfo() {
  console.log("\n=== Connection Info ===");
  const client = createClient(DB_CONFIG);
  const db = drizzle(client);

  const [vrow] = await db.execute(sql`SELECT version() AS version`);
  console.log(`CUBRID version: ${vrow.version}`);

  const [drow] = await db.execute(sql`SELECT database() AS db`);
  console.log(`Database: ${drow.db}`);

  const [urow] = await db.execute(sql`SELECT user() AS u`);
  console.log(`User: ${urow.u}`);

  await client.close();
}

async function multipleQueries() {
  console.log("\n=== Multiple Queries ===");
  const client = createClient(DB_CONFIG);
  const db = drizzle(client);

  const queries = [
    { label: "SELECT 1 AS a", query: sql`SELECT 1 AS a` },
    { label: "SELECT 'hello' AS b", query: sql`SELECT 'hello' AS b` },
    { label: "SELECT CURRENT_DATE AS today", query: sql`SELECT CURRENT_DATE AS today` },
  ];
  for (const { label, query } of queries) {
    const [row] = await db.execute(query);
    const value = Object.values(row)[0];
    console.log(`${label.padEnd(35)} → ${value}`);
  }

  await client.close();
}

async function main() {
  await basicConnection();
  await connectionInfo();
  await multipleQueries();
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: go/cubrid-go/01_connect.go
//go:build ignore

// 01_connect.go — Connecting to CUBRID with cubrid-go.
//
// Demonstrates:
// - Basic connection
// - Running a simple query
// - Connection metadata (version, database, user)
// - Reusing one connection for multiple queries

package main

import (
	"database/sql"
	"fmt"
	"log"

	_ "github.com/cubrid-labs/cubrid-go"
)

const dsn = "cubrid://dba:@localhost:33000/testdb"

func basicConnection(db *sql.DB) {
	fmt.Println("=== Basic Connection ===")

	var result int
	if err := db.QueryRow("SELECT 1 + 1 AS result").Scan(&result); err != nil {
		log.Fatal(err)
	}

	fmt.Printf("1 + 1 = %d\n", result)
}

func connectionInfo(db *sql.DB) {
	fmt.Println("\n=== Connection Info ===")

	var version string
	if err := db.QueryRow("SELECT version() AS version").Scan(&version); err != nil {
		log.Fatal(err)
	}
	fmt.Printf("CUBRID version: %s\n", version)

	var database string
	if err := db.QueryRow("SELECT database() AS db").Scan(&database); err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Database: %s\n", database)

	var user string
	if err := db.QueryRow("SELECT user() AS u").Scan(&user); err != nil {
		log.Fatal(err)
	}
	fmt.Printf("User: %s\n", user)
}

func multipleQueries(db *sql.DB) {
	fmt.Println("\n=== Multiple Queries ===")

	if _, err := db.Exec("DROP TABLE IF EXISTS cookbook_connect_demo"); err != nil {
		log.Fatal(err)
	}
	if _, err := db.Exec("CREATE TABLE cookbook_connect_demo (id INT, note VARCHAR(50))"); err != nil {
		log.Fatal(err)
	}
	if _, err := db.Exec("DROP TABLE cookbook_connect_demo"); err != nil {
		log.Fatal(err)
	}

	type queryExample struct {
		sql   string
	}

	queries := []queryExample{
		{sql: "SELECT 1 AS a"},
		{sql: "SELECT 'hello' AS b"},
		{sql: "SELECT CURRENT_DATE AS today"},
	}

	for _, q := range queries {
		var value any
		if err := db.QueryRow(q.sql).Scan(&value); err != nil {
			log.Fatal(err)
		}
		fmt.Printf("%-35s -> %v\n", q.sql, value)
	}
}

func main() {
	db, err := sql.Open("cubrid", dsn)
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	if err := db.Ping(); err != nil {
		log.Fatal(err)
	}

	basicConnection(db)
	connectionInfo(db)
	multipleQueries(db)
}
